pallets/offchain-demo/src/lib.rs

```diff
@@ -171,35 +202,72 @@ impl<T: Trait> Module<T> {
 		}
 	}
 
-	fn fetch_n_parse(_block_number: T::BlockNumber) -> Result<(), Error<T>> {
+	/// Check if we have fetched github info before. If yes, we use the cached version that is
+	///   stored in off-chain worker storage `storage`. If no, we fetch the remote info and then
+	///   write the info into the storage for future retrieval.
+	fn fetch_if_needed() -> Result<(), Error<T>> {
+
+		// Start off by creating a reference to Local Storage value.
+		// Since the local storage is common for all offchain workers, it's a good practice
+		// to prepend our entry with the pallet name.
+		let storage = StorageValueRef::persistent(b"offchain-demo::gh-info");
+
+		// The local storage is persisted and shared between runs of the offchain workers,
+		// and offchain workers may run concurrently. We can use the `mutate` function, to
+		// write a storage entry in an atomic fashion.
+		//
+		// It has a similar API as `StorageValue` that offer `get`, `set`, `mutate`.
+		// If we are using a get-check-set access pattern, we likely want to use `mutate` to access
+		// the storage in one go.
+		//
+		// Ref: https://substrate.dev/rustdocs/v2.0.0-alpha.6/sp_runtime/offchain/storage/struct.StorageValueRef.html
+		let res = storage.mutate(|store: Option<Option<GithubInfo>>| {
+			match store {
+				// info existed, returning the value
+				Some(Some(info)) => {
+					debug::info!("Using cached gh-info.");
+					Ok(info)
+				},
+				// info not existed, so we remote fetch (and parse the JSON)
+				_ => Self::fetch_n_parse(),
```

Comment:
This might be a bit sub-optimal. Mutation works by getting the value first and then comparing with previous value when writing. You are starting an external (blocking) HTTP request and that increases the chance of:
1. Mutliple OCW doing the same requests at the same time.
2. All except for one failing to write the storage.

A better approach is to:
1. Have OCW acquire a lock (mutate some storage entry) to perform the side effect (external call).
2. Then only do the call if acquiring the lock was successful.

This prevents doing the same request multiple times, but adds some extra complexity. I think it would be a good idea to have a `StorageLock` helper in substrate that would simplify this, basically:
```
let guard = StorageLock::acquire(max_duration_in_blocks);
/// do external calls
guard.release(): (or simply `Drop`)
```
The `max_duration_in_blocks` allows re-trying (we consider the lock "stale") in case the external call fails or is unable to complete within a reasonable timeout.

That said - the code here is correct, just doing the same external HTTP requests is a bit wasteful and doesn't show the `mutate` pattern correctly. If you want to keep this, I'd suggest just using `storage.set` instead of mutating, especially given that all external calls should end up with exactly the same results, so multiple writes won't cause issues.

Reply:
Reading your feedback, together with storage not shared across nodes, want to clarify something with you:

1. Currently there is only one ocw run per node, no more. When you say multiple OCWs, I assume you mean ocw in multiple nodes. What I don't understand about *having OCW acquire a lock* is, if ocw-storage is local and only one ocw per node. It doesn't matter if it is locked or not. To me, the storage lock is meaningful only when either multiple ocws per node, or storage is shared across nodes. Is this understanding correct?

2. Assuming there is lock in the ocw-storage, I assume when the closure function is run in `mutate`, that ocw has acquire a lock to run the closure, and other ocws (either in other nodes, or other ocw processes in the same node) are waiting outside. So there will only one ocw run the `fetch_n_parse()`. Once this is done and the ocw process returned, other ocw enter, see the value is set, and so skip the fetch. Is this understanding correct?

Reply:
We don't care about OCWs running on other nodes, OCW is something that is inherently local, there are no means to know or communicate with other nodes.

When I say multiple OCW runs I mean for instance OCW@Block 4 & OCW@Block 5. Workers run asynchronously and there might be a case where you have the same OCW (from the same pallet) running a the same time, but for a different block.

> To me, the storage lock is meaningful only when either multiple ocws per node, or storage is shared across nodes. Is this understanding correct?

Yup, that's correct. Storage is not shared, but as mentioned earlier we can have ocws for different blocks running at the same time.

> Once this is done and the ocw process returned, other ocw enter, see the value is set, and so skip the fetch. Is this understanding correct?

Yes. Instead of waiting you can also early-exit the workers if they fail to acquire the lock, cause that indicates that some other worker is already running.